### PR TITLE
Feature/848/compress output data

### LINKED ISSUE #848: Compress output data
# Feature request

The generated `cc.json` files might become quite big depending on the size of the project. As such, it would be great to minimize the file size and to compress them. It should be relatively straight forward to first implement something like this with an option and to later on switch the default as semver-major.

The data is very similar and repetitive. It is therefore likely that the file size would be a minimized to a maximum of 1/10 using a fast compression algorithm.

All tools handling the files should be able to handle compressed and uncompressed data.

### PR DESCRIPTION
**Pull request by JaiJoPa**
# Compress output data

closes #848

## Description

When creating a cc.json file, the user is now able to instead get a compressed cc.json.gz-File, which can be loaded into the visualization.
To compress the output data, just add a -c flag to the importer/parser command line.

## Definition of Done

A task/pull request will not be considered to be complete until all these items can be checked off.

- [ ] **All** requirements mentioned in the issue are implemented
- [ ] Does match the Code of Conduct and the Contribution file
- [ ] Task has its own **GitHub issue** (something it is solving)
  - [ ] Issue number is **included in the commit messages** for traceability
- [ ] **Update the README.md** with any changes/additions made
- [ ] **Update the CHANGELOG.md** with any changes/additions made
- [ ] **Enough test coverage to ensure that uncovered changes do not break functionality**
- [ ] **All tests pass**
- [ ] **Descriptive pull request text**, answering:
  - What problem/issue are you fixing?
  - What does this PR implement and how?
- [ ] **Assign your PR to someone** for a code review
  - This person _will be contacted **first**_ if a bug is introduced into `master`
- [ ] **Manual testing** did not fail

### CODE REVIEW

File: visualization/app/codeCharta/ui/fileChooser/fileChooser.component.ts [hunk 1]
```diff
@@ -26,23 +26,33 @@ export class FileChooserController {
...
+				const isCompressed = file.name.endsWith(".gz")
 				const reader = new FileReader()
 				reader.onloadstart = () => {
 					this.storeService.dispatch(setIsLoadingFile(true))
 				}
 				reader.onload = event => {
-					this.setNewData(file.name, (<any>event.target).result)
+					if (isCompressed) {
+						const pako = require("pako")
+						content = pako.ungzip((<any>event.target).result, { to: "string" })
```
Comment (BridgeAR) [inline visualization/app/codeCharta/ui/fileChooser/fileChooser.component.ts:40]: Please use the Node.js native unzip function https://nodejs.org/api/zlib.html. It is faster and does not require to use any extra library.

File: visualization/app/codeCharta/ui/fileChooser/fileChooser.component.ts [hunk 2]
```diff
@@ -26,23 +26,35 @@ export class FileChooserController {
...
 				const reader = new FileReader()
 				reader.onloadstart = () => {
 					this.storeService.dispatch(setIsLoadingFile(true))
 				}
 				reader.onload = event => {
-					this.setNewData(file.name, (<any>event.target).result)
+					if (isCompressed) {
+						const zlib = require("zlib")
+
+						const buffer = new Buffer((<any>event.target).result, "UTF-8")
```
Comment (BridgeAR) [inline visualization/app/codeCharta/ui/fileChooser/fileChooser.component.ts:41]: This should cause a deprecation warning while using it? Using `new Buffer()` could result in security vulnerabilities and should therefore definitely not be used. Instead, please use `Buffer.from()` as described 
here: https://nodejs.org/api/buffer.html

Comment (JaiJoPa) [reply visualization/app/codeCharta/ui/fileChooser/fileChooser.component.ts]: Oh you're right! It's deprecated. I changed it :)

File: visualization/app/codeCharta/ui/fileChooser/fileChooser.component.ts [hunk 3]
```diff
@@ -26,23 +26,34 @@ export class FileChooserController {
...
 				const reader = new FileReader()
 				reader.onloadstart = () => {
 					this.storeService.dispatch(setIsLoadingFile(true))
 				}
 				reader.onload = event => {
-					this.setNewData(file.name, (<any>event.target).result)
+					if (isCompressed) {
+						const zlib = require("zlib")
+
+						content = zlib.unzipSync(Buffer.from((<any>event.target).result))
```
Comment (BridgeAR) [inline visualization/app/codeCharta/ui/fileChooser/fileChooser.component.ts:41]: The sync function is going to slow down the process if there's more than a single file loaded. The reason is that it's a blocking operation. So no other task is able to run until this is done. Do you know if there are frequently multiple files?

**Pull request merged**